### PR TITLE
Add Full Rank Approximation

### PR DESCRIPTION
**Pull request by Sayam753**
This PR adds Full Rank ADVI interface. It is ready to review. Just to experiment with the API, I have created a [gist](https://gist.github.com/Sayam753/23592188b951bdeb53029eb0c4f4f2c3) inspired from @ColCarroll's [notebook](https://nbviewer.jupyter.org/gist/ColCarroll/d673a3af7169bd713bcbdb9445d4a543). I look forward to include mixture comparisons as well.

- [X] Add Full Rank ADVI
- [X] Add docstrings
- [X] Add tests
- [X] Change variable flattening to model flattening
- [X] Add Interval transformation
- [X] Rerun quick start notebook
- [X] Add deterministic callbacks
- [X] Add tests for deterministic callbacks

### CODE REVIEW

**Comment by fonnesbeck** [comment]
Just testing this now. Not sure what to make of this:

image

Should be 10K loss values, correct?

**Comment by Sayam753** [comment]
Hi @fonnesbeck 
Thanks for checking this out. The implementation of Full Rank ADVI(especially the model flattening part) is wrong. I am working on this! Soon, this will be resolved.

**Comment by fonnesbeck** [comment]
Mean field ADVI does not run at all. Fails with:

`ValueError: Dimensions must be equal, but are 4 and 1086 for '{{node monte_carlo_variational_loss/expectation/JointDistributionSequential/log_prob/add_1}} = AddV2[T=DT_FLOAT](monte_carlo_variational_loss/expectation/JointDistributionSequential/log_prob/add, monte_carlo_variational_loss/expectation/JointDistributionSequential/log_prob/Normal_1/log_prob/sub)' with input shapes: [1,4], [1,1086].`

**Comment by Sayam753** [comment]
Seems like a shape issue. Can you share a reproducible code snippet?
I suspect sort of adjusting `sample_size` parameter to `fit` function.

File: pymc4/variational/util.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        flattened_tensor = [tf.reshape(var, shape=[-1]) for var in self.free_rvs.values()]
+        return tf.concat(flattened_tensor, axis=0)
+
+    def split(self, flatten_tensor: tf.Tensor) -> Dict[str, tf.Tensor]:
+        """Split view of parameters used to calculate log probability."""
+        flat_state = dict()
+        for param in self.free_rvs:
+            _, slc, shape, dtype = self.by_name[param]
+            flat_state[param] = tf.cast(tf.reshape(flatten_tensor[slc], shape), dtype)
+        return flat_state
```
Comment (ferrine) [inline pymc4/variational/util.py:42]: need a function flat->dict

File: pymc4/variational/approximations.py [hunk 2]
```diff
@@ -66,13 +65,15 @@ def vectorized_logpfn(*q_samples):
...
-        """Flattened view of the variational parameters."""
-        pass
-
-    def sample(self, n):
+    def sample(self, n: int) -> az.InferenceData:
         """Generate samples from posterior distribution."""
-        q_samples = dict(zip(self.unobserved_keys, self.approx.sample(n)))
+        samples = self.approx.sample(n)
+        q_samples = dict()
+        for param in self.unobserved_keys:
```
Comment (ferrine) [inline pymc4/variational/approximations.py:72]: better be implemented in ordering

File: tests/test_variational.py [hunk 3]
```diff
@@ -31,6 +32,10 @@ def model():
     "ADVI": {
         "method": pm.MeanField, 
         "fit_kwargs": {}
+    },
```
Comment (ferrine) [inline tests/test_variational.py:35]: we need tests to make sure all making sense kwargs work properly (`sample_size`, etc)

**Comment by fonnesbeck** [comment]
Is this supposed to be working now? I'm getting the following:

image

It was running with the original PR, though not getting good results. Now it quits after a handful of iterations with NaNs.

**Comment by ferrine** [comment]
There are some concerns about transformations not being stable

**Comment by fonnesbeck** [comment]
Same behavior for mean-field ADVI

**Comment by Sayam753** [comment]
Yes. Some transformations were missing for Bounded Distributions. Highlighting the same, I had opened an issue #283 some time back. Commit d61431b adds the remaining transformations. Before this commit, passing `validate_args=True` to each distribution while doing VI, leads to various constraint errors from TFP. 

But I do not feel like its a good approach to manually add transformations to respective distributions. I tried to make it more general with a single class - 

```python
class Interval(BackwardTransform):
    name = "interval"

    def __init__(self, lower_limit, upper_limit):
        transform = tf.cond(
            tf.math.is_inf(lower_limit),
            lambda: tf.cond(
                tf.math.is_inf(upper_limit),
                lambda: tfb.Identity(),
                lambda: tfb.Chain([tfb.Shift(upper_limit), tfb.Scale(-1), tfb.Exp()]),  # upper - exp(x)
            ),
            lambda: tf.cond(
                tf.math.is_inf(upper_limit),
                lambda: tfb.Chain([tfb.Shift(lower_limit), tfb.Exp()]),  # exp(x) + lower
                lambda: tfb.Sigmoid(low=lower_limit, high=upper_limit),  # interval
            ),
        )
        super().__init__(transform)
```
But this leads to many eager execution issues for lambda functions. Please suggest if there is a way to improve existing approach.

**Comment by ferrine** [comment]
@fonnesbeck, what is park bias model?

**Comment by fonnesbeck** [comment]
@ferrine its a relatively large model that I was using to test ADVI, ported over from PyMC3. Its got a hierarchical component as well as Gaussian proccesses. Samples very slowly, so a good candidate for ADVI.

**Comment by Sayam753** [comment]
Hi @fonnesbeck 
The model seems interesting. Can you experiment again in your free time with the current state of PR? Issues related to transformations has been resolved.
Thanks

**Comment by fonnesbeck** [comment]
With the current PR, I am getting the following failure:

```
---------------------------------------------------------------------------
InvalidArgumentError                      Traceback (most recent call last)
<ipython-input-29-6a1ed8631a36> in <module>()
      1 m = park_bias_model()
      2 # trace = pm.sample(m, num_samples=100, burn_in=2000, xla=True)
----> 3 approx = pm.fit(m, num_steps=20000, method='fullrank_advi')

6 frames
/usr/local/lib/python3.6/dist-packages/pymc4/variational/approximations.py in fit(model, method, num_steps, sample_size, random_seed, optimizer, **kwargs)
    246         return losses
    247 
--> 248     return ADVIFit(inference, run_approximation())

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/def_function.py in __call__(self, *args, **kwds)
    778       else:
    779         compiler = "nonXla"
--> 780         result = self._call(*args, **kwds)
    781 
    782       new_tracing_count = self._get_tracing_count()

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/def_function.py in _call(self, *args, **kwds)
    844               *args, **kwds)
    845       # If we did not create any variables the trace we have is good enough.
--> 846       return self._concrete_stateful_fn._filtered_call(canon_args, canon_kwds)  # pylint: disable=protected-access
    847 
    848     def fn_with_cond(*inner_args, **inner_kwds):

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/function.py in _filtered_call(self, args, kwargs, cancellation_manager)
   1845                            resource_variable_ops.BaseResourceVariable))],
   1846         captured_inputs=self.captured_inputs,
-> 1847         cancellation_manager=cancellation_manager)
   1848 
   1849   def _call_flat(self, args, captured_inputs, cancellation_manager=None):

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/function.py in _call_flat(self, args, captured_inputs, cancellation_manager)
   1921       # No tape is watching; skip to running the function.
   1922       return self._build_call_outputs(self._inference_function.call(
-> 1923           ctx, args, cancellation_manager=cancellation_manager))
   1924     forward_backward = self._select_forward_and_backward_functions(
   1925         args,

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/function.py in call(self, ctx, args, cancellation_manager)
    548               inputs=args,
    549               attrs=attrs,
--> 550               ctx=ctx)
    551         else:
    552           outputs = execute.execute_with_cancellation(

/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/execute.py in quick_execute(op_name, num_outputs, inputs, attrs, ctx, name)
     58     ctx.ensure_initialized()
     59     tensors = pywrap_tfe.TFE_Py_Execute(ctx._handle, device_name, op_name,
---> 60                                         inputs, attrs, num_outputs)
     61   except core._NotOkStatusException as e:
     62     if name is not None:

InvalidArgumentError: 2 root error(s) found.
  (0) Invalid argument:  Cholesky decomposition was not successful. The input might not be valid.
	 [[{{node fit_surrogate_posterior/while/body/_22/fit_surrogate_posterior/while/StatefulPartitionedCall/monte_carlo_variational_loss/expectation/loop_body/PartitionedCall/pfor/PartitionedCall/Cholesky/pfor/Cholesky}}]]
	 [[fit_surrogate_posterior/while/body/_22/fit_surrogate_posterior/while/StatefulPartitionedCall/Adam/Adam/AssignAddVariableOp/_51]]
  (1) Invalid argument:  Cholesky decomposition was not successful. The input might not be valid.
	 [[{{node fit_surrogate_posterior/while/body/_22/fit_surrogate_posterior/while/StatefulPartitionedCall/monte_carlo_variational_loss/expectation/loop_body/PartitionedCall/pfor/PartitionedCall/Cholesky/pfor/Cholesky}}]]
0 successful operations.
0 derived errors ignored. [Op:__inference_run_approximation_207085]

Function call stack:
run_approximation -> run_approximation
```

**Comment by fonnesbeck** [comment]
Can you rebase to current master so that all the covariance functions are available?

File: tests/test_variational.py (hunk 3 above)
Comment (Sayam753) [reply tests/test_variational.py:35]: Resolved in ccd626c.

File: pymc4/variational/approximations.py (hunk 2 above)
Comment (Sayam753) [reply pymc4/variational/approximations.py]: Resolved in ccd626c.

File: pymc4/variational/util.py (hunk 1 above)
Comment (Sayam753) [reply pymc4/variational/util.py]: Resolved in ccd626c.

File: pymc4/variational/approximations.py [hunk 4]
```diff
@@ -42,25 +41,38 @@ def __init__(self, model: Optional[Model] = None, random_seed: Optional[int] = N
...
         @tf.function(autograph=False)
         def logpfn(*values):
             split_view = self.order.split(values[0])
             _, st = flow.evaluate_meta_model(self.model, values=split_view)
             return st.collect_log_prob()
 
-        def vectorize_logp_function(logpfn):
-            def vectorized_logpfn(*q_samples):
-                return tf.vectorized_map(lambda samples: logpfn(*samples), q_samples)
+        @tf.function(autograph=False)
```
Comment (ferrine) [inline pymc4/variational/approximations.py:56]: Let's also add a test to make sure it works

Comment (Sayam753) [reply pymc4/variational/approximations.py:56]: Resolved in 6189ba1.

File: pymc4/variational/approximations.py [hunk 5]
```diff
@@ -105,64 +103,67 @@ class MeanField(Approximation):
...
+    and Blei, D. M. (2016). Automatic Differentiation Variational
+    Inference. arXiv preprint arXiv:1603.00788.
+    """
 
     def _build_posterior(self):
-        raise NotImplementedError
+        flattened_shape = self.order.size
+        dtype = dtype_util.common_dtype(
+            self.state.all_unobserved_values.values(), dtype_hint=tf.float64
+        )
```
Comment (Sayam753) [inline pymc4/variational/approximations.py:138]: Figuring out reasons of Cholesky Decomposition failure, this [comment](https://github.com/ICL-SML/Doubly-Stochastic-DGP/issues/9#issuecomment-371167725) has been very helpful. Here are my few observations experimenting with VI and Gaussian Processes - 
1. High learning rates (order of 1e-2) leads to cholesky decomposition errors.
2. Both Mean Field and Full Rank ADVI also lead to decomposition error with `tf.float32` even if we add a small WhiteNoise to the diagonal. But testing ADVI with `tf.float64` leads to greater numerical stability of covariance matrix. So, I changed the `dtype_hint` to use `tf.float64`.

**Comment by ferrine** [comment]
and I also leave this experiment by @Sayam753 
image

**Comment by twiecki** [comment]
This is a major feature, congrats @Sayam753!

@ferrine should we open an issue to keep track of the issue?

**Comment by twiecki** [comment]
In another PR I would add full rank to the example NB, which also still reads "Full Rank ADVI - Coming Soon".

**Pull request merged**